pkg/util/ephemeral/label_matchers.go

```diff
@@ -0,0 +1,128 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package ephemeral
+
+import (
+	"fmt"
+	"strings"
+
+	amlabels "github.com/prometheus/alertmanager/pkg/labels"
+
+	"github.com/grafana/mimir/pkg/mimirpb"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"gopkg.in/yaml.v3"
+)
+
+// LabelMatchers configures matchers based on which series get marked as ephemeral.
+type LabelMatchers struct {
+	source []string
+	config []matcherSet
+	string string
+}
+
+// matcherSet is like alertmanager's labels.Matchers but for Prometheus' labels.Matcher slice
+type matcherSet []*labels.Matcher
+
+// matches checks whether all the matchers match the given label set.
+func (ms matcherSet) matches(lset []mimirpb.LabelAdapter) bool {
```

Comment:
interesting idea, until now I intentionally didn't sort the matchers to allow the user to configure the order in which they get checked, that way a user who wants to optimize the config could put `=` matchers before `=~` matchers. But I like your idea too, it could be more beneficial than relying on the user being smart, which often doesn't happen.

Reply:
I don't see this bit implemented, so I assume that we will optimize it in separate PR.

Reply:
ah yes, sorry for not being more explicit. i think it's a nice idea for an optimization, but for now i'd prefer to focus on implementing the functionality, later we can optimize it.